.github/workflows/run_mmperf.yml

```diff
@@ -0,0 +1,106 @@
+      runner-group: ${{ steps.configure.outputs.runner-group }}
+      write-caches: ${{ steps.configure.outputs.write-caches }}
+    steps:
+      - name: "Checking out repository"
+        uses: actions/checkout@7884fcad6b5d53d10323aee724dc68d8b9096a2e # v2
+        with:
+          # We need the parent commit to do a diff
+          fetch-depth: 2
+      - name: "Configuring CI options"
+        id: configure
+        run: |
+          # Just informative logging. There should only be two commits in the
+          # history here, but limiting the depth helps when copying from a local
+          # repo instead of using checkout, e.g. with
+          # https://github.com/nektos/act where there will be more.
+          git log --oneline --graph --max-count=3
+          ./build_tools/github_actions/configure_ci.py
+
+  benchmark_cpu:
+    needs: setup
+    if: needs.setup.outputs.should-run == 'true'
+    runs-on:
+      - self-hosted  # must come first
+      - runner-group=${{ needs.setup.outputs.runner-group }}
+      - environment=${{ needs.setup.outputs.runner-env }}
+      - cpu
+      - os-family=Linux
+    env:
+      MMPERF_RESULTS_DIR: mmperf-results
+      GCS_UPLOAD_DIR: gs://mmperf-benchmark-artifacts
+    steps:
+      - name: "Checking out repository"
+        uses: actions/checkout@7884fcad6b5d53d10323aee724dc68d8b9096a2e # v2
+        with:
+          submodules: true
+      - name: "Running mmperf on CPU"
+        run: |
+          mkdir ${MMPERF_RESULTS_DIR}
+          ./build_tools/github_actions/docker_run.sh \
+            gcr.io/iree-oss/mmperf@sha256:e563ead2db6f7c28a4fb815fe14444d38e40c49ba5b37b0d6dba741825570632 \
+          ./build_tools/benchmarks/mmperf/run_mmperf.sh "${MMPERF_RESULTS_DIR}" "cpu"
+      - name: "Uploading artifacts"
+        run: |
+          export DIR_NAME=`ls -lat ${MMPERF_RESULTS_DIR}/latest | awk '{print $NF}'`
+          gcloud alpha storage cp "${MMPERF_RESULTS_DIR}/${DIR_NAME}/**" "${GCS_UPLOAD_DIR}/cpu/${DIR_NAME}/"
+
+  benchmark_cuda:
+    needs: setup
+    if: needs.setup.outputs.should-run == 'true'
+    runs-on:
+      - self-hosted  # must come first
+      - runner-group=${{ needs.setup.outputs.runner-group }}
+      - environment=${{ needs.setup.outputs.runner-env }}
+      - gpu
+      - os-family=Linux
+    env:
+      MMPERF_RESULTS_DIR: mmperf-results
+      GCS_UPLOAD_DIR: gs://mmperf-benchmark-artifacts
+    steps:
+      - name: "Checking out repository"
+        uses: actions/checkout@7884fcad6b5d53d10323aee724dc68d8b9096a2e # v2
+        with:
+          submodules: true
+      - name: "Running mmperf on GPU"
+        run: |
+          mkdir ${MMPERF_RESULTS_DIR}
+          ./build_tools/github_actions/docker_run.sh \
+            --gpus all \
+            gcr.io/iree-oss/mmperf@sha256:e563ead2db6f7c28a4fb815fe14444d38e40c49ba5b37b0d6dba741825570632 \
+          ./build_tools/benchmarks/mmperf/run_mmperf.sh "${MMPERF_RESULTS_DIR}" "cuda"
```

Comment:
Time on GPU machines is more expensive than time on CPU machines, so we try to build on bulky CPU machines then run benchmarks on GPU machines. You could also consider using prebuilt compiler binaries, either from other CI runs (hard), or from the python releases (easier)

Looking through the logs on https://github.com/iree-org/iree/actions/runs/3373643809/jobs/5598425810 to see how much time is spent building vs running benchmarks:
```
Tue, 01 Nov 2022 23:41:11 GMT
-- Build files have been written to: /work/mmperf-build/matmul
Tue, 01 Nov 2022 23:41:11 GMT
[34/40] cd /work/mmperf-build/matmul-iree && /usr/bin/cmake --build .

Wed, 02 Nov 2022 00:09:48 GMT
[4892/4894] Linking CXX executable /work/mmperf-build/matmul/matmul_ireecuda_384x384x32

Wed, 02 Nov 2022 00:10:06 GMT
+ python3 mmperf/mmperf.py ./mmperf-build/matmul/ mmperf-results

Wed, 02 Nov 2022 00:20:34 GMT
ireecuda {'path': PosixPath('/work/mmperf-build/matmul/matmul_ireecuda_22201x32x288'), 'size': (22201, 32, 288), 'speed': 1311.5205297249136}
```

Looks like 54 minutes total, with ~30 minutes spent building the IREE compiler and other binaries and ~10 minutes spent running benchmarks

(also note that there are some warnings worth fixing in those logs)

Reply:
Good point. It would be ideal to use the Python releases but there's a fair amount of work in getting mmperf to use the Python releases. The benchmark code is in C++ and uses the IREE C API. I can look into it if this becomes a large bottleneck.

Reply:
It's fine to mix C runtime code with Python compiler binaries. The runtime has minimal dependencies and builds quickly (1-4 minutes), while the compiler pulls in large parts of LLVM and can take over 20 minutes to build even on a large machine.

Reply:
I've created an issue on mmperf to update this: https://github.com/mmperf/mmperf/issues/96

Reply:
Yeah an hour on GPU machines is going to cause issues. We need to not be building on the GPU machines. It looks like we spend a full 15 minutes installing things also. Can we dockerize that part? Overall, I see 45 minutes spent setting up and 10 minutes spent actually running. That's not really workable

Reply:
I made the following changes:
* Moved the git clone and venv setup to the Dockerfile.
* Due to permissions (root vs non-root) and CMake, we still need to build mmperf every time. I tried moving the GPU build step onto a CPU machine but this caused segmentation faults. Currently it's building on a GPU runner.

Reply:
Can you elaborate on the segmentation faults? It's troubling that something would be *unbuildable* without a GPU around

Reply:
The GPU build on a CPU machine did succeed but when running the benchmarks on a GPU machine, it would seg fault in the middle of running a model.

Reply:
> The GPU build on a CPU machine did succeed but when running the benchmarks on a GPU machine, it would seg fault in the middle of running a model.

That sounds like a serious bug, and not something to work around or ignore. It makes little sense in many deployment scenarios to compile and run on the same machine. Please file an issue.

Reply:
The seg faults may not be caused by building on a CPU machine. I'm still seeing seg faults after combining the build and run steps on the same machine. I'll need to debug this further.

Reply:
Fixed. The docker run command was missing the "--gpus all". We can now build the GPU image on a CPU machine. PTAL.